crates/matrix-sdk/src/notification_settings.rs

```diff
@@ -0,0 +1,782 @@
+            .await
+            .map_err(|_| NotificationSettingsError::UnableToSavePushRules)?;
+        Ok(())
+    }
+
+    /// Restore the default push rule for a given room
+    ///
+    /// Owner's account data will not be updated
+    pub fn restore_room_default_push_rule(
+        &self,
+        room_id: &String,
+        ruleset: &mut Ruleset,
+    ) -> Result<(), NotificationSettingsError> {
+        // Only if we have a custom mode for this room
+        if self.get_user_defined_room_notification_mode(room_id, ruleset).is_some() {
+            self.delete_user_defined_room_notification_mode(room_id, ruleset)?;
+        }
+        Ok(())
+    }
+
+    /// Sets a notification mode for a given room
+    ///
+    /// Owner's account data will not be updated
+    pub fn set_room_notification_mode(
+        &self,
+        room_id: &String,
+        mode: RoomNotificationMode,
+        ruleset: &mut Ruleset,
+    ) -> Result<(), NotificationSettingsError> {
+        match mode {
+            RoomNotificationMode::AllMessages => self.insert_notify_room_rule(room_id, ruleset),
+            RoomNotificationMode::MentionsAndKeywordsOnly => {
+                self.insert_mention_and_keywords_room_rule(room_id, ruleset)
+            }
+            RoomNotificationMode::Mute => self.insert_mute_room_rule(room_id, ruleset),
+        }
+    }
+
+    /// Gets the user defined push notification mode for a given room id
+    pub fn get_user_defined_room_notification_mode(
+        &self,
+        room_id: &String,
+        ruleset: &Ruleset,
+    ) -> Option<RoomNotificationMode> {
+        // Search for an enabled Override
+        if let Some(rule) = ruleset.override_.iter().find(|x| x.enabled) {
+            // without a Notify action
+            if !rule.actions.iter().any(|x| matches!(x, Action::Notify)) {
+                // with a condition of type `EventMatch` for this room_id
+                if rule.conditions.iter().any(|x| match x {
+                    PushCondition::EventMatch { key, pattern } => {
+                        key == "room_id" && *pattern == *room_id
+                    }
+                    _ => false,
+                }) {
+                    return Some(RoomNotificationMode::Mute);
+                }
+            }
+        }
+
+        // Search for an enabled Room rule where rule_id is the room_id
+        if let Some(rule) = ruleset.room.iter().find(|x| x.enabled && x.rule_id == *room_id) {
+            // if this rule contains a Notify action
+            if rule.actions.iter().any(|x| matches!(x, Action::Notify)) {
+                return Some(RoomNotificationMode::AllMessages);
+            }
+            // if this rule does not contain a Notify action
+            if !rule.actions.iter().any(|x| matches!(x, Action::Notify)) {
+                return Some(RoomNotificationMode::MentionsAndKeywordsOnly);
+            }
+            return Some(RoomNotificationMode::Mute);
```

Comment:
If I'm correct the 2 `if` should be replaced with `if/else`, we can remove this last return "Mute"